gcputil/credentials.go

```diff
@@ -50,6 +53,127 @@ type GcpCredentials struct {
+	// External Account fields
+	Audience              string   `json:"audience"`
+	ServiceAccountEmail   string   `json:"service_account_email"`
+	SubjectTokenType      string   `json:"subject_token_type"`
+	TokenURL              string   `json:"token_url"`
+	Scopes                []string `json:"scopes"`
+	WorkloadIdentityToken string   `json:"workload_identity_token"`
+}
+
+func (c *ExternalAccountCredential) TokenSource(ctx context.Context) (oauth2.TokenSource, error) {
+	ts := tokenSource{
+		ctx:    ctx,
+		config: c,
+	}
+	return oauth2.ReuseTokenSource(nil, ts), nil
+}
+
+func (c *ExternalAccountCredential) GetCredentials(ctx context.Context) (*google.Credentials, error) {
+	ts, err := c.TokenSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &google.Credentials{
+		TokenSource: ts,
+	}, nil
+}
+
+// tokenSource is the source that handles external credentials. It is used to retrieve Tokens.
+type tokenSource struct {
+	ctx    context.Context
+	config *ExternalAccountCredential
+}
+
+// Token allows tokenSource to conform to the oauth2.TokenSource interface.
+func (ts tokenSource) Token() (*oauth2.Token, error) {
+	// Exchange Vault's Identity Token for a federated STS Token
+	stsToken, err := ts.obtainSTSToken()
+	if err != nil {
+		return nil, err
+	}
+
+	// Exchange federated token for IAM Credential Token for the Service Account
+	saCredential, err := ts.obtainSACredential(stsToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return saCredential, nil
+}
+
+func (ts tokenSource) obtainSTSToken() (*oauth2.Token, error) {
+	// This STS Token Exchange is modeled after Google's oauth2 library
+	// For reference, please visit the following
+	// https://github.com/golang/oauth2/blob/master/google/internal/stsexchange/sts_exchange.go
+	stsRequest := STSTokenExchangeRequest{
+		GrantType:          "urn:ietf:params:oauth:grant-type:token-exchange",
+		Audience:           ts.config.Audience,
+		Scope:              ts.config.Scopes,
+		RequestedTokenType: "urn:ietf:params:oauth:token-type:access_token",
+		// plugin identity token fetched over system view
+		SubjectToken:     ts.config.WorkloadIdentityToken,
+		SubjectTokenType: ts.config.SubjectTokenType,
+	}
+	stsResp, err := ExchangeSTSToken(ts.ctx, ts.config.TokenURL, &stsRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	accessToken := &oauth2.Token{
+		AccessToken: stsResp.AccessToken,
+		TokenType:   stsResp.TokenType,
+	}
+	if stsResp.ExpiresIn < 0 {
+		return nil, fmt.Errorf("sts/google: got invalid expiry from security token service")
+	} else if stsResp.ExpiresIn >= 0 {
+		accessToken.Expiry = time.Now().Add(time.Duration(stsResp.ExpiresIn) * time.Second)
+	}
+
+	if stsResp.RefreshToken != "" {
+		accessToken.RefreshToken = stsResp.RefreshToken
+	}
+
+	return accessToken, nil
+}
+
+func (ts tokenSource) obtainSACredential(stsToken *oauth2.Token) (*oauth2.Token, error) {
+	// make a cURL request to the Service Account Access Token endpoint
```

Comment:
nit: cURL is a specific command line tool but here we are using the `go-cleanhttp` library
```suggestion
	// make a request to the Service Account Access Token endpoint
```